sdk/Sdk.Generators/ExtensionStartupRunnerGenerator.Parser.cs

```diff
@@ -0,0 +1,134 @@
+                IAssemblySymbol assemblySymbol, CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var extensionStartupAttribute = GetExtensionStartupAttributeData(assemblySymbol, compilation);
+                if (extensionStartupAttribute is null)
+                {
+                    return default;
+                }
+
+                // WorkerExtensionStartupAttribute has a constructor with one param, the type of startup implementation class.
+                var firstConstructorParam = extensionStartupAttribute.ConstructorArguments[0];
+                if (firstConstructorParam.Value is not INamedTypeSymbol startupTypeNamedTypeSymbol)
+                {
+                    return default;
+                }
+
+                return CreateStartupTypeInfoFromType(startupTypeNamedTypeSymbol, compilation);
+            }
+
+            private static StartupTypeInfo CreateStartupTypeInfoFromType(INamedTypeSymbol namedTypeSymbol,
+                Compilation compilation)
+            {
+
+                var builder = ImmutableArray.CreateBuilder<Diagnostic>();
+
+                // Check for any diagnostics we need to report about this type.                
+
+                // 1) Check public parameterless constructor exist for the type.
+                var constructorExist = namedTypeSymbol.InstanceConstructors
+                    .Any(c => c.Parameters.Length == 0 &&
+                              c.DeclaredAccessibility == Accessibility.Public);
+                if (!constructorExist)
+                {
+                    var diagnostic = Diagnostic.Create(DiagnosticDescriptors.ConstructorMissing, Location.None,
+                        namedTypeSymbol.ToDisplayString());
+                    builder.Add(diagnostic);
+                }
+
+                // 2) Check the extension startup class implements WorkerExtensionStartup abstract class.
+                INamedTypeSymbol? startupBaseTypeSymbol = compilation.GetTypeByMetadataName(StartupBaseClassName);
+                
+                if (namedTypeSymbol.BaseType == null ||
+                    !namedTypeSymbol.BaseType.Equals(startupBaseTypeSymbol, SymbolEqualityComparer.Default))
+                {
+                    var diagnostic = Diagnostic.Create(DiagnosticDescriptors.IncorrectBaseType, Location.None,
+                        namedTypeSymbol.ToDisplayString(), StartupBaseClassName);
+                    builder.Add(diagnostic);
+                }
+                
+                return new StartupTypeInfo(namedTypeSymbol.ToDisplayString(), builder.ToImmutable());
+            }
+        }
+
+        /// <summary>
+        /// A type to represent the data about the extension startup types we want to generate code for.
+        /// </summary>
+        internal struct StartupTypeInfo
```

Comment:
I think you can mark this struct `readonly`. This will let the compiler avoid defensive copies (good for perf).